tlv/record_type.go

```diff
@@ -0,0 +1,59 @@
+package tlv
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"golang.org/x/exp/constraints"
+)
+
+// Primitive is a type constraint that capture the set of "primitive" types,
+// which are the built in stdlib types, and type defs of those types
+type Primitive interface {
+	constraints.Integer | ~[]byte | ~[32]byte | ~[33]byte | ~bool |
+		~*btcec.PublicKey | ~[64]byte
+}
+
+// RecordT is a high-order type makes it easy to encode known "primitive" types
+// as TLV records.
+type RecordT[T Record | Primitive] struct {
```

Comment:
I find the use of `Record` as a type constraint a bit sus here. I combed through some golang documentation and it seems like there isn't an explicit mention (at least one that was easy to find) of the fact that you can use struct definitions as interface constraints.

After reading [this](https://go.dev/blog/intro-generics) with a little creative inference I'm roughly understanding this to mean that any type that either **is** a `Record` or a type that **implements** `Primitive` may be used here. But I don't think that's actually what we want imo. What we want is any type that can be deterministically serialized and deserialized via the wire definitions, after which we annotate it with a type number and arrive at a fully defined tlv record. I **believe** this is more akin to the `RecordProducer` interface, though not completely since the RecordProducer interface currently takes responsibility for the type number annotation.

I think what we need is to define a new interface that all of the `Primitive` types implement that is only responsible for understanding how to serialize itself sans tlv type number. So in this case it would know how serialize and then lengh-prefix. From there the `Record[T]` type could be constrained on this interface basically saying: "as long as you know how to pack yourself down to a sequence of bytes, I will provide a slot for annotating you with the proper type number".

I had started to make this change in my branch on the wire types for DynComms but held off because I wanted to build off of this. If you're amenable to this slightly different way of thinking about the problem, I could make the changes and PR them into this branch to save you the time.

Reply:
> I find the use of Record as a type constraint a bit sus here. I combed through some golang documentation and it seems like there isn't an explicit mention (at least one that was easy to find) of the fact that you can use struct definitions as interface constraints.

Type constraints can be any type. This is how you declare things like: 
```go
type Integer interface {
     uint | uint64
}
```

> What we want is any type that can be deterministically serialized and deserialized via the wire definitions, after which we annotate it with a type number and arrive at a fully defined tlv record. I believe this is more akin to the RecordProducer interface, though not completely since the RecordProducer interface currently takes responsibility for the type number annotation.

So this isn't yet at the level of serialization. It's accepting anything that is a primitive type, or a record. Serialization is still done via the `tlv.Stream` struct. This isn't about defining a new `Encode/Decode` method on all top-level structs. That's a larger overhaul to the package that can be done later, when I'm seeking something minimal here to make the package a bit easier to work with. To carry out such an overhaul, we'd need to do away with the current encode/decode _functions_ eg (`tlv.EUint8`), and instead make them into methods (abstracted via a new interface), with those changes reverberating through the entire project. 

As mentioned below, I just want to make the smallest change possible here, as we _can_ rewrite the entire package, but I think we have higher priority tasks that aren't necessarily blocked on having a 100% idealized TLV package.

Reply:
I'm not suggesting doing away with the existing record type where we drop the functions in there. Rather I'm talking about revamping how we do the equivalent of `RecordProducer`. Since tlv type numbers are not strictly a property of the data type but rather of the use context of the field, it doesn't make sense to make RecordProducer responsible for annotating the data with the type number. Instead, what you present here is much better suited to managing the type annotations, the only requirement that your `Record[T]` struct has to do its job is to make sure the T it houses can be written to the wire and length prefixed, something that can be delegated to T itself via interfaces.